lib/views/app.js

```diff
@@ -0,0 +1,145 @@
+
+/**
+ * Dependencies
+ */
+
+var ReactNative = require('react-native');
+var debug = require('../debug')('App');
+var Scanner = require('../scanner');
+var ListView = require('./list');
+var React = require('react');
+
+var {
+  LayoutAnimation,
+  StyleSheet,
+  Alert,
+  View
+} = ReactNative;
+
+/**
+ * The length of time to scan for.
+ *
+ * @type {Number}
+ */
+const SCAN_LENGTH = 8000; // 8secs
+
+/**
+ * The length of time between scans.
+ *
+ * @type {Number}
+ */
+const SCAN_INTERVAL = 30000; // 30secs
+
+class AppView extends React.Component {
+  constructor(props) {
+    super(props);
+
+    // initial state
+    this.state = {
+      awaitingFirstScan: true,
+      scanning: false,
+      items: []
+    };
+
+    // bind callback context
+    this.scan = this.scan.bind(this);
+
+    // create scanner (accepts test mock)
+    this.scanner = props.scanner || new Scanner(); // eslint-disable-line
+
+    this.scanner.on('update', this.onScannerUpdate.bind(this));
+    this.scanner.on('networkerror', this.onScannerNetworkError.bind(this));
+
+    // Finding something before react-native is
+    // fully initialized can cause a crash:
+    // https://github.com/walmartreact/react-native-orientation-listener/issues/8
+    // TODO: This should be fixed inside NetworkScanner.java
+    setTimeout(this.scan, 100);
+  }
+
+  scan(options={}) {
+    if (this.state.scanning) return;
+    debug('scanning ...');
+
+    clearTimeout(this.nextScanTimeout);
+
+    this.setState({
+      awaitingFirstScan: false,
+      scanning: true
+    });
+
+    if (options.clear) this.scanner.clear();
+
+    this.scanner.scan(SCAN_LENGTH).then(() => {
+      this.nextScanTimeout = setTimeout(this.scan, SCAN_INTERVAL);
+      this.setState({ scanning: false });
+      debug('scan complete');
+    });
+  }
+
+  render() {
+    debug('render', this.state.items);
+    var loading = this.state.awaitingFirstScan || this.state.scanning;
+
+    return (
+      <View style={[styles.root, this.props.style]}>
+        <ListView
+          style={styles.list}
+          items={this.state.items}
+          loading={loading}
+          onRefresh={this.onRefresh.bind(this)}/>
+      </View>
+    )
+  }
+
+  onRefresh() {
+    debug('on refresh');
+    this.scan({ clear: true });
+  }
+
+  onScannerUpdate(items) {
+    debug('on scanner update', items);
+    LayoutAnimation.configureNext(LayoutAnimation.Presets.easeInEaseOut);
+    this.setState({ items: items });
+  }
+
+  onScannerNetworkError() {
+    debug('on network error');
```

Comment:
No, the `.scan()` will eventually finish and switch the `scanning` state back to `false. Although this could be improved if we called `scanner.stop()` when a network error happens.